1111-i386-txt-Add-Intel-TXT-ACM-module-support.patch

```diff
@@ -0,0 +1,612 @@
++    }
++
++  if ( n_fit_in_acm ( hdr, id_entries_off, sizeof(struct grub_txt_acm_processor_id),
++		     proc_id_list->count ) == NULL )
++    {
++      grub_error (GRUB_ERR_OUT_OF_RANGE, N_("ACM chipset ID entries out of bounds"));
++      return NULL;
++    }
++
++  return proc_id_list;
++}
++
++static int
++is_acmod (const void *acmod_base, grub_uint32_t acmod_size,
++         grub_uint8_t *type_out)
++{
++  struct grub_txt_acm_header *acm_hdr = (struct grub_txt_acm_header *)acmod_base;
++  struct grub_txt_acm_info_table *info_table;
++  grub_uint32_t size_from_hdr;
++
++  /* First check size */
++  if ( acmod_size < sizeof (struct grub_txt_acm_header) )
++    return 0;
++
++  /* Then check overflow */
++  if ( grub_mul (acm_hdr->size, 4, &size_from_hdr) )
++    return 0;
++
++  /* Then check size equivalency */
++  if ( acmod_size != size_from_hdr )
++    return 0;
++
++  /* Then check type, sub-type and vendor */
++  if ( (acm_hdr->module_type != GRUB_TXT_ACM_MODULE_TYPE) ||
++       (acm_hdr->module_sub_type != GRUB_TXT_ACM_MODULE_SUB_TYPE_TXT_ACM) ||
++       (acm_hdr->module_vendor != GRUB_TXT_ACM_MODULE_VENDOR_INTEL) )
++    return 0;
++
++  info_table = get_acmod_info_table (acm_hdr);
++  if ( info_table == NULL )
++    return 0;
++
++  /* Check if ACM UUID is present */
++  if ( grub_memcmp (&(info_table->uuid), GRUB_TXT_ACM_UUID, 16) )
++    return 0;
++
++  /*
++   * TXT specification doesn't give clear mapping of info table size to version,
++   * so just warn if the size is different than expected but try to use it
++   * anyway. info_table_offset() macro does enough testing to not read outside
++   * of info table.
++   */
++  if ( info_table->length < sizeof(struct grub_txt_acm_info_table) )
++    grub_dprintf ("slaunch", "Info table size (%x) smaller than expected (%x)\n",
++		  info_table->length, sizeof(struct grub_txt_acm_info_table));
++
++  if ( type_out )
++    *type_out = info_table_offset (info_table, chipset_acm_type);
++
++  return 1;
++}
++
++static struct grub_txt_acm_header*
++get_bios_sinit (void *sinit_region_base)
++{
++  grub_uint8_t *txt_heap = grub_txt_get_heap ();
++  struct grub_txt_bios_data *bios_data = grub_txt_bios_data_start (txt_heap);
++  struct grub_txt_acm_header *bios_sinit;
++  grub_uint32_t tmp;
++
++  if ( sinit_region_base == NULL )
++    return NULL;
++
++  if ( bios_data->bios_sinit_size == 0 )
++    return NULL;
++
++  /* Check if ACM crosses 4G */
++  if ( grub_add ( (unsigned long)sinit_region_base, bios_data->bios_sinit_size,
++                  &tmp) )
++    return NULL;
++
++  /* BIOS has loaded an SINIT module, so verify that it is valid */
++  grub_dprintf ("slaunch", "BIOS has already loaded an SINIT module\n");
++
++  bios_sinit = (struct grub_txt_acm_header *)sinit_region_base;
++
++  /* Is it a valid SINIT module? */
++  if ( !grub_txt_is_sinit_acmod (sinit_region_base, bios_data->bios_sinit_size) ||
++       !grub_txt_acmod_match_platform (bios_sinit) )
++    return NULL;
```

Comment:
Considering the above `grub_dprintf`, perhaps we should log
```c
grub_dprintf("slaunch", "BIOS SINIT module did not pass sanity checks");
```

Reply:
Fair point, added.